File: mac-server/Mac2Droid/Protocol/FrameHeader.swift
import Foundation

// MARK: - Handshake Packet
/// Initial handshake sent when client connects (24 bytes)
struct M2DHandshake {
    let magic: [UInt8]      // 4 bytes: "M2D\0"
    let version: UInt32     // 4 bytes: protocol version
    let codec: M2DCodec     // 4 bytes: codec type
    let width: UInt32       // 4 bytes: video width
    let height: UInt32      // 4 bytes: video height
    let frameRate: UInt32   // 4 bytes: target FPS

    init(width: Int, height: Int, frameRate: Int, codec: M2DCodec = .h264) {
        self.magic = M2DProtocol.magic
        self.version = M2DProtocol.version
        self.codec = codec
        self.width = UInt32(width)
        self.height = UInt32(height)
        self.frameRate = UInt32(frameRate)
    }

    /// Serialize handshake to Data (24 bytes)
    func serialize() -> Data {
        var data = Data(capacity: M2DProtocol.handshakeSize)

        // Magic (4 bytes)
        data.append(contentsOf: magic)

        // Version (4 bytes, big-endian)
        data.append(contentsOf: version.bigEndianBytes)

        // Codec (4 bytes, big-endian)
        data.append(contentsOf: codec.rawValue.bigEndianBytes)

        // Width (4 bytes, big-endian)
        data.append(contentsOf: width.bigEndianBytes)

        // Height (4 bytes, big-endian)
        data.append(contentsOf: height.bigEndianBytes)

        // Frame rate (4 bytes, big-endian)
        data.append(contentsOf: frameRate.bigEndianBytes)

        return data
    }
}

// MARK: - Frame Header
/// Header for each video frame packet (12 bytes)
struct M2DFrameHeader {
    let flags: M2DFrameFlags   // 1 byte: frame flags
    let reserved: UInt8        // 1 byte: reserved for future use
    let pts: UInt64            // 6 bytes: presentation timestamp (microseconds)
    let payloadSize: UInt32    // 4 bytes: size of NAL data following header

    init(flags: M2DFrameFlags, pts: UInt64, payloadSize: Int) {
        self.flags = flags
        self.reserved = 0
        self.pts = pts
        self.payloadSize = UInt32(payloadSize)
    }

    /// Create header for a video frame
    static func videoFrame(pts: UInt64, payloadSize: Int, isKeyframe: Bool, isConfig: Bool) -> M2DFrameHeader {
        var flags = M2DFrameFlags()
        if isKeyframe {
            flags.insert(.keyframe)
        }
        if isConfig {
            flags.insert(.config)
        }
        return M2DFrameHeader(flags: flags, pts: pts, payloadSize: payloadSize)
    }

    /// Create end-of-stream header
    static func endOfStream() -> M2DFrameHeader {
        return M2DFrameHeader(flags: .endOfStream, pts: 0, payloadSize: 0)
    }

    /// Serialize header to Data (12 bytes)
    func serialize() -> Data {
        var data = Data(capacity: M2DProtocol.frameHeaderSize)

        // Flags (1 byte)
        data.append(flags.rawValue)

        // Reserved (1 byte)
        data.append(reserved)

        // PTS (6 bytes, big-endian) - only use lower 48 bits
        let ptsBytes = pts.bigEndianBytes
        data.append(contentsOf: ptsBytes.suffix(6))

        // Payload size (4 bytes, big-endian)
        data.append(contentsOf: payloadSize.bigEndianBytes)

        return data
    }
}

// MARK: - Byte Conversion Extensions
extension UInt32 {
    var bigEndianBytes: [UInt8] {
        let value = self.bigEndian
        return withUnsafeBytes(of: value) { Array($0) }
    }
}

extension UInt64 {
    var bigEndianBytes: [UInt8] {
        let value = self.bigEndian
        return withUnsafeBytes(of: value) { Array($0) }
    }
}

extension Data {
    /// Read big-endian UInt32 from specified offset
    func readUInt32(at offset: Int) -> UInt32 {
        guard offset + 4 <= count else { return 0 }
        var value: UInt32 = 0
        _ = copyBytes(to: UnsafeMutableBufferPointer(start: &value, count: 1), from: offset..<(offset + 4))
        return UInt32(bigEndian: value)
    }

    /// Read big-endian 6-byte value as UInt64 from specified offset
    func readUInt48(at offset: Int) -> UInt64 {
        guard offset + 6 <= count else { return 0 }
        var bytes = [UInt8](repeating: 0, count: 8)
        copyBytes(to: &bytes[2], from: offset..<(offset + 6))
        return bytes.withUnsafeBytes { $0.load(as: UInt64.self).bigEndian }
    }
}

File: android-client/app/src/main/kotlin/com/mac2droid/protocol/FrameHeader.kt
package com.mac2droid.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Stream configuration from handshake
 */
data class StreamConfig(
    val version: Int,
    val codec: M2DCodec,
    val width: Int,
    val height: Int,
    val frameRate: Int
) {
    companion object {
        /**
         * Parse handshake packet (24 bytes)
         */
        fun parse(data: ByteArray): StreamConfig? {
            if (data.size < M2DProtocol.HANDSHAKE_SIZE) return null

            val buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)

            // Verify magic
            val magic = ByteArray(4)
            buffer.get(magic)
            if (!magic.contentEquals(M2DProtocol.MAGIC)) return null

            val version = buffer.int
            val codecValue = buffer.int
            val width = buffer.int
            val height = buffer.int
            val frameRate = buffer.int

            val codec = M2DCodec.fromValue(codecValue) ?: return null

            return StreamConfig(
                version = version,
                codec = codec,
                width = width,
                height = height,
                frameRate = frameRate
            )
        }
    }
}

/**
 * Frame header for each video packet
 */
data class FrameHeader(
    val isConfig: Boolean,
    val isKeyframe: Boolean,
    val isEndOfStream: Boolean,
    val pts: Long,          // Presentation timestamp in microseconds
    val payloadSize: Int
) {
    companion object {
        /**
         * Parse frame header (12 bytes)
         */
        fun parse(data: ByteArray): FrameHeader? {
            if (data.size < M2DProtocol.FRAME_HEADER_SIZE) return null

            val buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)

            // Flags (1 byte)
            val flags = buffer.get().toInt() and 0xFF
            val isConfig = (flags and M2DFrameFlags.CONFIG) != 0
            val isKeyframe = (flags and M2DFrameFlags.KEYFRAME) != 0
            val isEndOfStream = (flags and M2DFrameFlags.END_OF_STREAM) != 0

            // Reserved (1 byte)
            buffer.get()

            // PTS (6 bytes) - read as part of 8-byte long
            val ptsBytes = ByteArray(8)
            buffer.get(ptsBytes, 2, 6)
            val pts = ByteBuffer.wrap(ptsBytes).order(ByteOrder.BIG_ENDIAN).long

            // Payload size (4 bytes)
            val payloadSize = buffer.int

            return FrameHeader(
                isConfig = isConfig,
                isKeyframe = isKeyframe,
                isEndOfStream = isEndOfStream,
                pts = pts,
                payloadSize = payloadSize
            )
        }
    }
}

/**
 * NAL unit parser for Annex B format
 */
object NalParser {
    /**
     * Extract SPS and PPS from config data
     * Config data format: [start_code][SPS][start_code][PPS]
     */
    fun extractParameterSets(configData: ByteArray): Pair<ByteArray, ByteArray>? {
        val nalUnits = extractNalUnits(configData)
        if (nalUnits.size < 2) return null

        // First NAL should be SPS (type 7), second should be PPS (type 8)
        val sps = nalUnits.find { getNalType(it) == 7 } ?: return null
        val pps = nalUnits.find { getNalType(it) == 8 } ?: return null

        return Pair(sps, pps)
    }

    /**
     * Extract NAL units from Annex B format data
     */
    fun extractNalUnits(data: ByteArray): List<ByteArray> {
        val units = mutableListOf<ByteArray>()
        var start = 0

        while (start < data.size) {
            // Find start code
            val startCodePos = findStartCode(data, start)
            if (startCodePos < 0) break

            // Find next start code or end of data
            val nalStart = startCodePos + 4
            var nalEnd = data.size

            val nextStartCode = findStartCode(data, nalStart)
            if (nextStartCode >= 0) {
                nalEnd = nextStartCode
            }

            // Extract NAL unit
            if (nalEnd > nalStart) {
                units.add(data.copyOfRange(nalStart, nalEnd))
            }

            start = nalEnd
        }

        return units
    }

    /**
     * Find 4-byte start code position starting from offset
     */
    private fun findStartCode(data: ByteArray, offset: Int): Int {
        for (i in offset until data.size - 3) {
            if (data[i] == 0x00.toByte() &&
                data[i + 1] == 0x00.toByte() &&
                data[i + 2] == 0x00.toByte() &&
                data[i + 3] == 0x01.toByte()) {
                return i
            }
        }
        return -1
    }

    /**
     * Get NAL unit type from first byte
     */
    fun getNalType(nalUnit: ByteArray): Int {
        return if (nalUnit.isNotEmpty()) nalUnit[0].toInt() and 0x1F else -1
    }
}
